rust/otap-dataflow/crates/otap/src/otap_grpc/proxy.rs

```diff
@@ -0,0 +1,939 @@
+    stream.set_nodelay(tcp_nodelay)?;
+
+    let std_stream = stream.into_std()?;
+    let socket: Socket = std_stream.into();
+
+    // Apply TCP keepalive settings
+    if let Some(keepalive_time) = tcp_keepalive {
+        let mut keepalive = TcpKeepalive::new().with_time(keepalive_time);
+
+        if let Some(interval) = tcp_keepalive_interval {
+            keepalive = keepalive.with_interval(interval);
+        }
+
+        #[cfg(not(target_os = "windows"))]
+        if let Some(retries) = tcp_keepalive_retries {
+            keepalive = keepalive.with_retries(retries);
+        }
+
+        #[cfg(target_os = "windows")]
+        if tcp_keepalive_retries.is_some() {
+            otap_df_telemetry::otel_warn!(
+                "Proxy.KeepaliveRetriesIgnored",
+                platform = "windows",
+                message = "tcp_keepalive_retries is configured but ignored on Windows: TcpKeepalive::with_retries is not available on this platform"
+            );
+        }
+
+        socket.set_tcp_keepalive(&keepalive)?;
+    }
+
+    // Convert back to std TcpStream, then to tokio TcpStream
+    let std_stream: std::net::TcpStream = socket.into();
+    std_stream.set_nonblocking(true)?;
+    TcpStream::from_std(std_stream)
+}
+
+/// Establishes a TCP connection to a target, optionally through an HTTP CONNECT proxy.
+///
+/// This is intended for gRPC transports that manage TLS separately (e.g., tonic's
+/// `Endpoint` with `.tls_config(...)`).
+///
+/// Note: TCP socket options (nodelay, keepalive) are applied to the connection
+/// to the proxy server itself. Since the tunnel is just a byte stream over this
+/// connection, these settings effectively apply to the tunneled traffic as well.
+pub(crate) async fn connect_tcp_stream_with_proxy_config(
+    target_uri: &Uri,
+    proxy_config: &ProxyConfig,
+    tcp_nodelay: bool,
+    tcp_keepalive: Option<Duration>,
+    tcp_keepalive_interval: Option<Duration>,
+    tcp_keepalive_retries: Option<u32>,
+) -> Result<TcpStream, ProxyError> {
+    let scheme = target_uri.scheme_str().unwrap_or("http");
+    let host = target_uri
+        .host()
+        .ok_or_else(|| ProxyError::InvalidUri("missing host".to_string()))?;
+    let port = target_uri.port_u16().unwrap_or(match scheme {
+        "https" => 443,
+        _ => 80,
+    });
+
+    if let Some(proxy_url) = proxy_config.get_proxy_for_uri(target_uri) {
+        let (proxy_host, proxy_port, proxy_auth) = parse_proxy_url(proxy_url)?;
+
+        otap_df_telemetry::otel_debug!("Proxy.Connecting", host = proxy_host, port = proxy_port);
+        let stream = TcpStream::connect((proxy_host.as_str(), proxy_port))
+            .await
+            .map_err(|e| {
+                otap_df_telemetry::otel_warn!(
+                    "Proxy.ConnectFailed",
+                    host = proxy_host,
+                    port = proxy_port,
+                    error = e.to_string()
+                );
+                ProxyError::ProxyConnectionFailed(e)
```

Comment:
Not sure but depending of what is reported in io::Error we might here leak some sensitive data.